src/webui/api/torrentscontroller.cpp

```diff
@@ -852,16 +991,47 @@ void TorrentsController::addAction()
         }
     };
 
+
     bool partialSuccess = false;
-    for (QString url : asConst(urls.split(u'\n')))
+    for (QString url : urls)
     {
         url = url.trimmed();
         if (!url.isEmpty())
         {
+            auto iter = m_torrentSource.constFind(url);
+            if (iter == m_torrentSource.constEnd())
+                iter = m_torrentSource.constFind(FILE_SOURCE_PREFIX + url);
+
+            // check whether metadata has already been fetched for this url
+            if ((iter != m_torrentSource.constEnd()) && isMetadataDownloaded(iter.value()))
+            {
+                const BitTorrent::InfoHash &hash = iter.value();
+                const BitTorrent::TorrentDescriptor torrentDescr = m_torrentMetadata[hash];
```

Comment:
> I would put all this logic to some function `std::optional<TorrentDescriptor> getMetadata(const QString &source)` to avoid complicating main logic.

Agreed and updated.

> Also I would avoid double lookup (use of `isMetadataDownloaded(iter.value())` and subsequent use of `m_torrentMetadata[hash]`).

I'll look into this. It's slightly tricky given the many uses of `isMetadataDownloaded`.